### PR TITLE
feat: Historical Contractor Payments SDK (AI pilot experiment)

### PR DESCRIPTION
**Pull request by larchai**
## Context

This is the output of an **AI-assisted SDK component generation pilot** — not intended for production merge as-is. We're sharing it for eng commentary on pattern adherence, architectural decisions, and feasibility of this approach as a development accelerator.

**Background:** Historical Contractor Payments was descoped from the Contractor Payments SDK launch because mixing historical and future-dated payments in one flow caused UX confusion. This PR builds a standalone `HistoricalPaymentFlow` component for recording past-dated contractor payments.

## What's in this PR

A complete `HistoricalPaymentFlow` with a 3-step journey:

1. **Select a past payment date** — Date picker constrained to past dates, defaulting to yesterday
2. **Select contractors and specify amounts** — Same table/modal pattern as existing CreatePayment, minus payment method selector (always "Historical Payment")
3. **Review and submit** — Preview via API, submit with creationToken, success confirmation

**15 new files** under `src/components/Contractor/Payments/HistoricalPayments/` following all existing SDK patterns: robot3 state machine, container/presentation split, BaseComponent, ComponentsContext, Zod forms, i18n, DataView, DOMPurify, React Query hooks.

See the [README](src/components/Contractor/Payments/HistoricalPayments/README.md) for full details on how it was built, file inventory, and current status.

## How to test

1. Check out this branch
2. `npm install && npm run sdk-app`
3. Provision a demo company (or use existing)
4. Find **"Contractor.Payments.HistoricalPayments.HistoricalPaymentFlow"** in the sidebar
5. Walk through the full flow with live data

## What's done

- Full flow works end-to-end with live API data (preview + create)
- TypeScript, ESLint, Prettier, commitlint all pass
- 31/32 PRD acceptance criteria met (1 minor — date range is permissive, defers to API validation)
- Pattern consistency with existing Contractor Payments SDK verified across all 15 architectural patterns

## What's NOT done (would need before production)

- Unit tests (Vitest)
- Storybook stories
- Accessibility audit
- Design review
- Edge case handling (0 contractors, network errors, etc.)
- Integration into existing PaymentFlow (currently standalone only)

## What we're looking for

This is shared for **commentary, not merge approval**:

- **Pattern adherence** — Does the code follow SDK conventions correctly?
- **Architectural decisions** — Is the simplified 3-state machine right, or should preview be a separate state?
- **Gaps** — What edge cases or requirements did the AI miss?
- **AI pilot viability** — Is this approach viable for accelerating future SDK component development?

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/CreateHistoricalPayment.tsx [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+    </BaseComponent>
+  )
+}
+
+export const Root = ({ companyId, dictionary, onEvent }: CreateHistoricalPaymentProps) => {
+  useComponentDictionary(
+    'Contractor.Payments.HistoricalPayments.CreateHistoricalPayment',
+    dictionary,
+  )
+  const { t } = useTranslation('Contractor.Payments.HistoricalPayments.CreateHistoricalPayment')
```
Comment (serikjensen) [inline src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/CreateHistoricalPayment.tsx:51]: I think we would want to create a separate dedicated domain for this, `Contractor.HistoricalPayments` rather than nesting this many values on the translation key

File: src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/CreateHistoricalPayment.tsx [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+            total: acc.total + hourlyAmount + fixedWage + bonus + reimbursement,
+          }
+        },
+        { wage: 0, bonus: 0, reimbursement: 0, total: 0 },
+      ),
+    [virtualContractorPayments, contractors],
+  )
+
+  const formMethods = useForm<EditHistoricalPaymentFormValues>({
+    resolver: zodResolver(createEditHistoricalPaymentFormSchema()),
```
Comment (serikjensen) [inline src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/CreateHistoricalPayment.tsx:114]: We likely don't need a create function on the schema for this one

File: src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/CreateHistoricalPayment.tsx [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+            wage: acc.wage + fixedWage,
+            bonus: acc.bonus + bonus,
+            reimbursement: acc.reimbursement + reimbursement,
+            total: acc.total + hourlyAmount + fixedWage + bonus + reimbursement,
+          }
+        },
+        { wage: 0, bonus: 0, reimbursement: 0, total: 0 },
+      ),
+    [virtualContractorPayments, contractors],
+  )
```
Comment (serikjensen) [inline src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/CreateHistoricalPayment.tsx:111]: I think there's some reworking here that could take place from the mapping -> virtualization. i think we could update this to be driven by the modal submission rather than needing to memoize

File: src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/CreateHistoricalPayment.tsx [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+        wageType: contractor?.wageType || 'Hourly',
+        hours: Number(contractorPayment?.hours || '0'),
+        wage: Number(contractorPayment?.wage || '0'),
+        bonus: Number(contractorPayment?.bonus || '0'),
+        reimbursement: Number(contractorPayment?.reimbursement || '0'),
+        hourlyRate: Number(contractor?.hourlyRate || '0'),
+        contractorUuid: contractorUuid,
+      },
+      { keepDirty: false, keepValues: false },
+    )
```
Comment (serikjensen) [inline src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/CreateHistoricalPayment.tsx:171]: I think we can update this to conditionally render the modal so we get the form reset out of the box rather than trying to reset form state. There are some issues that can happen with forms that are tough to track down with the attempted manual reset

File: src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/CreateHistoricalPayment.tsx [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+              paymentMethod: HISTORICAL_PAYMENT_METHOD,
+              isTouched: hasAnyPayment,
+            }
+          : payment,
+      ),
+    )
+    const displayContractor = contractors.find(
+      contractor => contractor.uuid === data.contractorUuid,
+    )
+    const displayName = DOMPurify.sanitize(
```
Comment (serikjensen) [inline src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/CreateHistoricalPayment.tsx:202]: I believe we already have a first/last name utility that is using DOMPurify under the hood

File: src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/CreateHistoricalPayment.tsx [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+        },
+      },
+    }))
+    setIsModalOpen(false)
+    onEvent(componentEvents.CONTRACTOR_HISTORICAL_PAYMENT_UPDATE, data)
+  }
+
+  const onContinueToPreview = async () => {
+    await baseSubmitHandler(null, async () => {
+      const today = new Date().toISOString().split('T')[0]
```
Comment (serikjensen) [inline src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/CreateHistoricalPayment.tsx:232]: This should be able to use an existing date util

File: src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/CreateHistoricalPayment.tsx [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+          paymentDate={paymentDate}
+          onPaymentDateChange={setPaymentDate}
+          onSaveAndContinue={onContinueToPreview}
+          onEditContractor={onEditContractor}
+          totals={totals}
+          alerts={alerts}
+          isLoading={isCreatingContractorPaymentGroup || isPreviewingContractorPaymentGroup}
+        />
+      )}
+      <EditHistoricalPaymentPresentation
```
Comment (serikjensen) [inline src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/CreateHistoricalPayment.tsx:303]: i think the rec here would be to supply a unique key here that can be incremented after update or similar approach to unmount the form rather than needing to run the reset

File: src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/EditHistoricalPaymentPresentation.tsx [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    control: formMethods.control,
+  })
+  const reimbursement = useWatch<EditHistoricalPaymentFormValues, 'reimbursement'>({
+    name: 'reimbursement',
+    control: formMethods.control,
+  })
+  const hourlyRate = useWatch<EditHistoricalPaymentFormValues, 'hourlyRate'>({
+    name: 'hourlyRate',
+    control: formMethods.control,
+  })
```
Comment (serikjensen) [inline src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/EditHistoricalPaymentPresentation.tsx:55]: Nit: My rec here would be to consider firing this onChange for the associated inputs. i worry a little about the perf overhead of watching every input. we'd have to check with the live form to see if it was a perceivable issue

File: src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/EditHistoricalPaymentPresentation.tsx [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+                  label={t('reimbursementLabel')}
+                  format="currency"
+                />
+              </Grid>
+            </Flex>
+          </Flex>
+        </Form>
+      </FormProvider>
+    </Modal>
+  )
```
Comment (serikjensen) [inline src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/EditHistoricalPaymentPresentation.tsx:143]: Def want to check this business logic for these conditionally rendered fields

File: src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/EditHistoricalPaymentFormSchema.ts [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+export const EditHistoricalPaymentFormSchema = z.object({
+  wageType: z.enum(['Hourly', 'Fixed']),
+  hours: z.number().nonnegative().max(20000).optional(),
+  wage: z.number().nonnegative().optional(),
+  bonus: z.number().nonnegative().optional(),
+  reimbursement: z.number().nonnegative().optional(),
+  hourlyRate: z.number().nonnegative().optional(),
+  contractorUuid: z.string(),
+})
```
Comment (serikjensen) [inline src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/EditHistoricalPaymentFormSchema.ts:11]: there are conditionally rendered fields but a flat schema. i assume that works because most of these are optional, but still would want to stress test to see if we can get invalid combinations and if the schema should be properly refining

File: src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/helpers.ts [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+export const getMaxHistoricalDate = (): string => {
+  const today = new Date()
+  return today.toISOString().split('T')[0] || ''
+}
+
+export const getMinHistoricalDate = (): string => {
+  const twoYearsAgo = new Date()
+  twoYearsAgo.setFullYear(twoYearsAgo.getFullYear() - 2)
+  return twoYearsAgo.toISOString().split('T')[0] || ''
+}
```
Comment (serikjensen) [inline src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/helpers.ts:30]: I think this may need to use existing date utils to avoid issues with UTC time in different timezones

File: src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/HistoricalPreviewPresentation.tsx [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+          acc.wageAmount += Number(contractor.wage || '0')
+          acc.bonusAmount += Number(contractor.bonus || '0')
+          acc.reimbursementAmount += Number(contractor.reimbursement || '0')
+          acc.totalAmount += Number(contractor.wageTotal || '0')
+          return acc
+        },
+        { wageAmount: 0, bonusAmount: 0, reimbursementAmount: 0, totalAmount: 0 },
+      ),
+    [contractorPaymentGroup.contractorPayments],
+  )
```
Comment (serikjensen) [inline src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/HistoricalPreviewPresentation.tsx:57]: These totals calculations are getting repeated, would recommend breaking this into a shared util for contractor payments

File: src/components/Contractor/Payments/HistoricalPayments/types.ts [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+
+export type InternalAlert = {
+  type: 'error' | 'info' | 'success'
+  title: string
+  content?: ReactNode
+  onDismiss?: () => void
+  translationParams?: Record<string, unknown>
+  onAction?: () => void
+  actionLabel?: string
+}
```
Comment (serikjensen) [inline src/components/Contractor/Payments/HistoricalPayments/types.ts:16]: Not sure why this is duplicating the alert props?

File: src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/EditHistoricalPaymentPresentation.tsx [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+            {t('cancelCta')}
+          </Button>
+          <Button
+            variant="primary"
+            type="submit"
+            form={formId}
+            onClick={() => formMethods.handleSubmit(onSubmit)}
+          >
+            {t('saveCta')}
+          </Button>
```
Comment (serikjensen) [inline src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/EditHistoricalPaymentPresentation.tsx:79]: We should either remove `type="submit"` _or_ update to associate the button action with the form id

Also doesn't this button need a loading/pending state?

File: src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/CreateHistoricalPayment.tsx [hunk 15]
```diff
@@ -0,0 +1,313 @@
...
+          contractorPayments={virtualContractorPayments}
+          paymentDate={paymentDate}
+          onPaymentDateChange={setPaymentDate}
+          onSaveAndContinue={onContinueToPreview}
+          onEditContractor={onEditContractor}
+          totals={totals}
+          alerts={alerts}
+          isLoading={isCreatingContractorPaymentGroup || isPreviewingContractorPaymentGroup}
+        />
+      )}
```
Comment (serikjensen) [inline src/components/Contractor/Payments/HistoricalPayments/CreateHistoricalPayment/CreateHistoricalPayment.tsx:302]: I think this would need more thought from the eng side of things on how to compose these components. Having CreateHistoricalPayment fork on the preview or the create historical payment is coupling these things too tightly. I think we'd likely want the preview as a dedicated component living as a separate entry in the state machine and then we would navigate to that preview based on an event fired from this component. As is, this component is doing too much

File: src/shared/constants.ts [hunk 16]
```diff
@@ -131,6 +131,16 @@ export const contractorPaymentEvents = {
   CONTRACTOR_PAYMENT_RFI_RESPOND: 'contractor/payments/rfi/respond',
 } as const
 
+export const contractorHistoricalPaymentEvents = {
+  CONTRACTOR_HISTORICAL_PAYMENT_CREATE: 'contractor/historicalPayments/create',
```
Comment (serikjensen) [inline src/shared/constants.ts:135]: This is never actually emitted and can be removed

**Review by serikjensen — COMMENTED** [review]
I think overall this captured a lot of the existing conventions for component composition. I think the biggest miss was the combination of the preview and create historical payment into the same component which makes these tangled in a way that partners wouldn't be able to consume individually and also has the CreateHistoricalPayment component doing too much. This is one in which having a baseline eng consult to organize the state machine and component composition as a first pass would be beneficial.

We could also consider trying to distill some of that into a create flow skill as something to watch out for when spinning up flows.

In any case, my other largest rec would be to decompose for submission of the state machine and then each of the components individually for easier review and verification, likely merging the state machine first to get the baseline organization that will impact all of the other downstream components.

Other issues I flagged that introduce subtleties that may not be immediately obvious
* Manual form reset - it's not always apparent that this isn't working until you are doing some combo of triggering validation issues, clearing them, running submit but I have had trouble in the past with getting react hook form to properly reset and generally prefer to actually unmount via conditional rendering or react key
* Button with onSubmit and onClick will fire twice and could lead to erroneous behavior that isn't always predictable
* Virtualization has memoization that has a deep object that would not throw on referential equality so i would def want to stress test that and maybe just consider a more event driven approach to that composition instead
* I worry about the conditional field rendering and if it has correctly captured the business logic. The flat schema that corresponds also is worth digging into more

Overall if the purpose of this exercise is to determine where our gaps are with LLM generation for larger flows I think that goes into a bigger question of how we work as an eng team which is obviously going to evolve over time. But the biggest shortcoming is on the composition of the block components and how those would interact with the flow which would have implications for hooks build out as well as how partners would use these components. I think there's value in having more eng collaboration up front to make sure the base infrastructure and component organization is in place even as AI fleshes out the implementation details to help remedy that. I also think at the current snapshot in time there's still value in working smaller and submitting PRs that are more digestable.

**Comment by serikjensen** [comment]
Closing this one as we will revisit historical payments with our updated design processes
